### ISSUE TITLE
Djvulibre stack corruption --- TC issues?

### ISSUE THREAD

**Issue by tigran123** [koreader/kindlepdfviewer#401]
I am seeing some serious problems (independent of my recent DjVu optimizations) with viewing DjVu files. Consistent crashes with core files pointing to stack corruption. Not reproducible in the emulator at all, only the ARM binary is bad. I am suspecting the TC and now am building on 32bit to see if it makes a difference and also will try building with the old good arm-kindle TC (but I'll have to revert changes to Makefile to match arm-kindle TC).

Update: replaced the word "random" with "consistent" :)

**Comment by tigran123** [comment]
The binary generated on 64 bit shows this stacktrace:

```
(gdb) bt
#0  0x0013d8c8 in ?? ()
Cannot access memory at address 0x0
#1  0x0013dae0 in ?? ()
Cannot access memory at address 0x0
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
```

The binary generated on 32 bit shows a different stracktrace:

```
(gdb) bt
#0  0x001238ac in lj_str_new ()
#1  0x00123f88 in lj_str_fromint ()
#2  0x00000064 in ?? ()
Backtrace stopped: frame did not save the PC
```

However, there is no guarantee that they both crashed in the same place (though what I was doing was exactly identical). You can reproduce the crash easily, just download this binary:

https://github.com/downloads/tigran123/libdjvu/libdjvu-test.djvu

and then goto page 15 (`G` 15) and press `Shift-X` to define the bbox. It will either crash immediately or allow you to set the top-left part and will crash on setting the bottom-right. Works fine in the emulator, crashes on Kindle 3.

**Comment by tigran123** [comment]
Update, the 28th September version works fine --- that was the time just before we switched to the new TC:

https://github.com/downloads/tigran123/kindlepdfviewer/kindlepdfviewer-496fcc0.zip

I'll try to narrow down to see which is the first build that introduced these crashes.

**Comment by tigran123** [comment]
Update: all the builds that I have made available work fine:

https://github.com/tigran123/kindlepdfviewer/downloads

The last build is dated 7th October and by then we have already switched to the new TC, so it doesn't look like a TC issue. We just need to check carefully what went into the master between the build `v2012.09-114-gcb25029` and now that broke DjVu.

Looks like a good idea to have regular daily builds --- helps to narrow down issues like this one.

**Comment by tigran123** [comment]
I didn't realize that Kindle has gdb installed :)
Using it I see yet another stacktrace (manually copied, as I don't know to do screendumps in myts):

```
propagatemark()
lj_gc_step()
lua_pushlstring()
push_captures()
str_find_aux()
lj_BC_FUNCC()
lua_pcall()
main()
```

**Comment by dpavlin** [comment]
FWIW, screenshots can be made using `dd if=/dev/fb0 of=/tmp/screen.fb0` and than use lua code which we have in KPV or https://github.com/dpavlin/k3libre/blob/master/fb2pgm.pl to convert it to image.

**Comment by tigran123** [comment]
Update: it can still be a TC issue. The 114 release (7 October kindlepdfviewer-v2012.09-114-gcb25029.zip from my Downloads) crashes in exactly the same way (I just didn't try hard enough, i.e. you have to press `Next` as well, not just `Shift-X` and `Shift-S`). So, now I am going back from 7 October into the past, testing each release thoroughly to see when the problem started to appear.

**Comment by tigran123** [comment]
Ok, the last build that I cannot break (and I tried very hard) is 99, i.e. this one:

https://github.com/downloads/tigran123/kindlepdfviewer/kindlepdfviewer-v2012.09-99-gea92f8c.zip

So, something went into master between `v2012.09-99-gea92f8c` and `v2012.09-114-gcb25029` that broke DjVu. Now `git diff` is our friend...

**Comment by tigran123** [comment]
Hmmm, the most serious thing that went into that release is #372. But I'll narrow down the commits one by one to be absolutely certain now.

**Comment by tigran123** [comment]
It is getting more interesting... I managed to cause the build 99 to crash as well, doing the same thing (only "more of it"). But the crash is very intriguing (@NiLuJe are you there?), look:

```
Program terminated with signal 4. Illegal instruction.
```

So, definitely looks like a TC issue, i.e. the ARM binary we generated contained instructions illegal for this particular processor. That is why there is no way to reproduce it on the emulator.

**Comment by tigran123** [comment]
The key question is whether I will be able to crash the build kindlepdfviewer-b6936f9.zip (1 Oct), i.e. the last build made with the old TC....

**Comment by tigran123** [comment]
No, I can't break the one built with the old TC, so it does look like TC issue as I originally suspected. To prove it I need to cause it to crash with the very first build that used the new TC. And then we'll need to figure out how to downgrade back to arm-kindle-linux-gnueabi --- it is going to be tricky as there were some useful changes to the Makefile that should still be kept even with the old TC....

**Comment by tigran123** [comment]
Btw, Kindle 3 is running Linux 2.6.26 and comparing kpdfview binary built with x-tools and the latest CS/MG I see:

```
$ file kpdfview*
kpdfview-MG:     ELF 32-bit LSB executable, ARM, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.16, stripped
kpdfview-xtools: ELF 32-bit LSB executable, ARM, version 1 (SYSV), dynamically linked (uses shared libs), for GNU/Linux 2.6.26, stripped
```

So, according to `file(1)` the xtools binary is more closely matching the Kindle's kernel, i.e. 2.6.26.

**Comment by hwhw** [comment]
I'm slowly getting back into things and maybe, it would be worthwhile to update luajit. It had lots of bug fixes and the latest beta release is - according to Mike Pall - about what will be finally the stable 2.0 release. I'll have a try.

Other than that, the error messages do look a bit as if a Lua string gets tampered with in C land.

**Comment by tigran123** [comment]
@hwhw interesting, because I've just reproduced one (only one!) of the "ARM-specific" crashes on the emulator, i.e. where on the ARM binary we get "signal 4, Illegal instruction" --- on the emulator we get the usual sig11 and backtrace looks like this:

```
(gdb) bt full
#0  0x00000000004fe210 in lua_pushnumber ()
No symbol table info available.
#1  0x0000000000544acc in walkTableOfContent (L=0x75766a642e3833, L@entry=0x41763378, r=<optimized out>, count=count@entry=0x7fff2669218c, depth=1, depth@entry=0)
    at djvu.c:140
        lista = 0x1a6ffd0
        length = 9
        counter = <optimized out>
        page_number = "0all_4"
#2  0x0000000000544e13 in getTableOfContent (L=0x41763378) at djvu.c:181
        doc = 0x4056d308
        r = <optimized out>
        count = 7
#3  0x0000000000510433 in lj_BC_FUNCC ()
No symbol table info available.
#4  0x00000000004ff360 in lua_pcall ()
No symbol table info available.
#5  0x000000000040765e in main (argc=<optimized out>, argv=0x7fff26692328) at kpdfview.c:74
        i = <optimized out>
(gdb) 
```

But this still leaves the potential TC issue, i.e. it is not proven yet one way or the other than the new TC is not responsible for some of the crashes (i.e. those that do look really ARM-specific) I am seeing. To reproduce the above just open this djvu file:

https://github.com/downloads/tigran123/libdjvu/toc-crash-test.djvu

and press `T`

**Comment by tigran123** [comment]
@hwhw also, I would definitely recommend to get to the bottom of these crashes _before_ upgrading luajit (but this is probably obvious anyway :)

**Comment by hwhw** [comment]
I have a feeling the recursive nature of walkTableOfContent() is blowing the available space on the Lua stack (which is limited!). There is an API to check available space on the Lua stack, but if handled carefully, that should not be needed. I'll dig into it now.

**Comment by tigran123** [comment]
Also, I should warn that the above stacktrace was obtained when running the emulator on 64bit Linux (and compiled also on the same) which is known to produce wrong stacktraces.

I am going to recompile it now on 32bit to make sure the stacktrace is accurate, i.e. we are not chasing a red herring :)
(although `T` command does suggest that the above stacktrace is accurate)

**Comment by hwhw** [comment]
hm, also, ignore my feeling mentioned above. stack usage seems to be fine in a first test.

**Comment by tigran123** [comment]
I get the same stacktrace on 32bit EMU:

```
(gdb) bt full
#0  walkTableOfContent (L=0x6370632e, L@entry=0x766a642e, r=<optimized out>, count=0x766a642e, 
    count@entry=0xbfcee1fc, depth=1, depth@entry=0) at djvu.c:140
        lista = 0x89f0075
        length = 9
        counter = <optimized out>
        page_number = "0all_4"
#1  0x0818ca2d in getTableOfContent (L=0xb76ae1c0) at djvu.c:181
        doc = 0x37313030
        r = <optimized out>
        count = 7
#2  0x08159307 in lj_BC_FUNCC ()
No symbol table info available.
#3  0x081488cd in lua_pcall ()
No symbol table info available.
#4  0x0804dc61 in main (argc=3, argv=0xbfcee324) at kpdfview.c:74
        i = <optimized out>
```

**Comment by hwhw** [comment]
Also, switching to newest luajit doesn't change this error.

**Comment by tigran123** [comment]
Ok, so we have a nice plain bug --- all that is left is just to fix it :)

**Comment by hwhw** [comment]
It's the strcpy. Unbounded strcpy is problematic, replacing it by a strncpy(...,6) works for me. However, there is no guarantee then that the last byte in page_number will be a '\0' anymore, so I guess it will need more space in general and a proper initialization with memset.

**Comment by tigran123** [comment]
I have just extracted djvu outline for the problematic book with `djvused -e print-outline book.djvu > book.outline` and I see that it has a lot of entries that are NOT numbers (and so using `lua_pushnumber()` would probably crash), look:

```
(bookmarks
 ("Contents"
  "#all_4160_to_00070.cpc0003.djvu" )
 ("Preface"
  "#all_4160_to_00070.cpc0005.djvu" )
 ("Part One - Preliminaries"
  "#all_4160_to_00070.cpc0017.djvu"
  ("1. Beginnings"
   "#all_4160_to_00070.cpc0019.djvu" )
  ("2. Brain and Reality"
   "#all_4160_to_00070.cpc0028.djvu" )
  ("3. Mathematics and Classical Reality"
   "#all_4160_to_00070.cpc0038.djvu" ) )
```

All those things starting with `#` should be target page numbers.

**Comment by tigran123** [comment]
I wonder how `djview4` copes with this --- somehow it jumps to the correct place when you click on those TOC entries...

**Comment by hwhw** [comment]
I'm not into what makes a proper "bookmark" in djvus, but strtol will in any case return a number. It's a wrong number in these cases, that's for sure. An option I could think of would be to return the string as-is to lualand, but this would not match the API presented by e.g. pdf.c and cre.cpp. However, cre.cpp already extends this by another field, "xpointer". Maybe we should do the same for page references that do not constitute "numbers"?

**Comment by tigran123** [comment]
The stacktrace pointed to line 140 of `djvu.c`:

```
139:   while(counter < length-1) {
140:        lua_pushnumber(L, *count);
141:        lua_newtable(L);
```

So, I was thinking that the crash is related to the above call to `lua_pushnumber()` and not to the subsequent call dealing with `page_number`.

But yes, the `page_number` according to stacktrace is `0all_4` which is not a number, but this shouldn't crash because, as you rightly say, `strtol()` would only return a number.

**Comment by tigran123** [comment]
Ok, instead of much talking "in theory" I'll now put some debug `printf()` to understand what is going on :)

**Comment by hwhw** [comment]
or rather use "break djvu.c:146" or similar in gdb, and output what's interesting via "print <expression>". worked for me. as for the theory: strcpy will happily copy up to as many bytes as there are until the '\0' is seen. As you dumped above via djvused, the string is much longer than just "0all_4" -- so strcpy will overwrite memory following page_number.

segfaults are often caused at other points than where they manifest in the end. Since page_number is on the stack here, the stack gets mangled with badly. Which BTW would also explain the SIGILL on ARM, since it is well possible that a return address gets overwritten so it might jump back in the middle of nowhere.

**Comment by tigran123** [comment]
Ah, yes, that makes perfect sense. Let me see if that fixes it (i.e. prevents the crash). And then we'll think how to deal with these strange entries. Maybe we should just convert them to proper page numbers without leaving djvu.c. Also, if you view the file with djview4 and go to "Information" you will see that the real page numbers are also available, so we should be able to get at them properly (by "properly" I mean NOT by parsing those "#all_4160_to_00070.cpc0003.djvu" things we are given).

**Comment by hwhw** [comment]
I'll prepare a fix shortly, completely getting rid with the strcpy and page_number alltogether. Also, miniexp_to_str might return NULL, which isn't cared for currently either. Will take about another ten minutes here.

strtol will set errno when it couldn't parse the string at hand, so we should catch that case and do "the right thing" then, as you commented. I'll look into how djview4 does this.

**Comment by tigran123** [comment]
@hwhw Yes, I confirm that your analysis is correct, thank you --- with strncpy(...,6) we get the correct TOC tree but of course we can't jump anywhere because all the targets are invalid.

**Comment by tigran123** [comment]
Thank you, then I'll go to lunch now --- can't keep four females (inc a 4-months old one) waiting at the table for me for too long now :)

**Comment by hwhw** [comment]
Oh dear: http://djvu.git.sourceforge.net/git/gitweb.cgi?p=djvu/djview.git;a=blob;f=src/qdjview.cpp;h=545df52541ec5ab97f5e1adf3bc10f4cddfc5f5e;hb=HEAD#l3224

It's page names, not page numbers what we see. They have a complex syntax, in principle, they are hyperlinks and might also be relative (probably a seldom case in a TOC, if not even forbidden)....

**Comment by tigran123** [comment]
Hmmm, doing all those searches will probably slow down TOC retrieval (by a few milliseconds ;)
Fortunately, it is only initialized once.

**Comment by tigran123** [comment]
Btw, if you can get at page title (i, ii, iii, .... in the above djvu file) then it would be nice to show them in djvu-specific menu status at the bottom, i.e. in addition to "22/264 (4)" and also two flavours of goto: one for physical and and another logical numbers, e.g. `G` command to go to physical page number (as it is now) and new `Shift-G` to accept logical, i.e. page title like `viii`.

**Comment by hwhw** [comment]
I created a preliminary fix, as I now have to leave the keyboard again for today. It will still restrict everything to numerical references, will look into names in the coming days if nobody beats me to it. Feel free to test and merge.

**Comment by tigran123** [comment]
Ok, thank you!

**Comment by tigran123** [comment]
Ok, with this fix merged I will now retest everything to see if I still get mysterious crashes on ARM which I suspected to be related to using the latest CS/MG toolchain.

**Comment by tigran123** [comment]
Alas, the crash is still there (but totally unrelated to the above TOC crash, obviously). I will try to track this one down...

**Comment by tigran123** [comment]
The stacktrace does point to luajit:

```
(gdb) bt
#0  0x00109748 in gc_marktrace ()
#1  0x00109cf4 in gc_traverse_trace ()
#2  0x0010a378 in propagatemark ()
#3  0x0010ab64 in lj_gc_step ()
#4  0x0000001a in ?? ()
```

but I know of no easy way of upgrading to the latest luajit to see if this problem was fixed. But I can dig some more to see what is the cause of it...

Besides, it wouldn't explain why the problem didn't occur until we switched to the new TC. But we did change the way luajit is compiled...

Ah, I recall seeing `-Wl,-E` (or something like that) in one of @NiLuJe 's private commits (not in our master tree) with the comment saying "luajit sources tell that we should link with this switch if we build it statically". Maybe that is something to look into....

**Comment by hwhw** [comment]
I'll send a pull request soon. It's as simpel as doing a "git pull -f" in
the luajit-2.0 directory if you want to do it yourself. Also, in the main
Makefile the "distclean" for luajit must be changed to "clean". When this
has been done, you can commit this change with a "git commit luajit-2.0
Makefile" if everything works.

**Comment by tigran123** [comment]
Thanks, I will definitely try that!

Currently, I am backtracking to the state when we built with x-tools to make 100% sure that it worked there.

**Comment by hwhw** [comment]
On the other hand, it points to the memory management of LuaJIT, which could be indicating that it's just another case of memory corruption. Is it also happening when pressing "T" for that file?

**Comment by tigran123** [comment]
No, pressing `T` on that file is fine now. You think it is a memory corruption. I think it is the CS/MG TC generating invalid binary. My reason is --- I can't reproduce these crashes in the emulator, only on the real ARM.

The `T` crash was reproduced immediately in the emulator and it turned out to be a memory corruption bug.

**Comment by tigran123** [comment]
This is a very complex problem, so we have to be methodical. Let's see what we know:
1. The 1 Oct build seems to work fine: https://github.com/downloads/tigran123/kindlepdfviewer/kindlepdfviewer-b6936f9.zip
2. The current master crashes and so does 7 Oct build: https://github.com/downloads/tigran123/kindlepdfviewer/kindlepdfviewer-v2012.09-114-gcb25029.zip
3. The first build that uses the new TC is 3 Oct: https://github.com/downloads/tigran123/kindlepdfviewer/kindlepdfviewer-v2012.09-64-g6ea40b4.zip I am not yet able to cause it to crash.

What I need to do now is again to go through my list of builds and find out exactly the first one where the crashes appeared. But only after I am happy that the 3 Oct build is okay (this would prove that it is not due to the new TC).

**Comment by tigran123** [comment]
@hwhw Concerning LuaJIT I am a little bit confused...
Namely, what is it that controls which version of LuaJIT is picked up by "make fetchthirdparty"? According to .gitmodules it would just do `git clone http://luajit.org/git/luajit-2.0.git` which according to the instructions at http://luajit.org/download.html would pick up the latest code.

On the other hand, if I create a new directory and do `git clone http://luajit.org/git/luajit-2.0.git` into it and then run diff -urN between it and the kindlepdfviewer/luajit-2.0 version then there are LOTS of differences. So, there must be something in kindlepdfviewer which specifies which version of luajit-2.0 to checkout.

**Comment by tigran123** [comment]
Also, running `git pull -f` doesn't work:

```
$ pwd
/home/tigran/Kindle/kindlepdfviewer/luajit-2.0
$ git pull -f
You asked to pull from the remote '-f', but did not specify
a branch. Because this is not the default configured remote
for your current branch, you must specify a branch on the command line.
```

**Comment by tigran123** [comment]
Update: with the latest LuaJIT (just manually cloned and then replaced the luajit-2.0 directory) it still crashes but with a different stacktrace:

```
lj_str_new()
lj_str_fromnum()
lj_ff_tostring()
lua_pcall()
main()
```

Again, only on the ARM, not reproducible in the emulator. And on the ARM one has to turn a few pages (with `Shift-S` and back to `A`) and press `Menu` and `Vol+/-` a few times. I still don't have the minimal sequence of actions that causes the crash.

**Comment by tigran123** [comment]
Ok, @hwhw may be right, it does look more and more like a memory corruption issue, because I can't force either 64 (3 Oct) or 73 (4 Oct) builds to crash and these were definitely built with the new TC. So, I need to bisect it problem to find which commit(s) introduced the bug.

**Comment by tigran123** [comment]
Ok, we are back to where we were 9 hours ago, i.e. the last non-crashing build is 99 (the reason it crashed before was due to TOC memory corruption which is a separate issue fixed by @hwhw today). So, something between builds 99 and 114 is causing the crashes, i.e. between the labels `v2012.09-99-gea92f8c` and `v2012.09-114-gcb25029`.

**Comment by tigran123** [comment]
It appears that my changes to `rendertext.lua` are causing these mysterious ARM-only crashes, but I will double-check to make 100% sure. The crashes are sometimes hard to reproduce so I may have marked some commit on my list as "good" when really it is "bad". But I will definitely get to the bottom of this...

**Comment by tigran123** [comment]
Ah, I forgot the local in front of `buffer_width = buffer:getWidth()` in `renderUtf8Text()`, so it overwrote whatever global `buffer_width` there may be! Hmm, this is just a guess from looking at the code. I will test it now.

**Comment by tigran123** [comment]
Yes, that seems to be it --- I can't break it anymore... I can't believe that such a tiny thing like missing `local` can waste one's whole day chasing the wild geese and catching red herrings :)

**Comment by tigran123** [comment]
Hmmm, but we don't have any global variable called `buffer_width` and yet adding that local does make a huge difference --- I can't break DjVu anymore! :)

**Comment by tigran123** [comment]
I am leaving this open in case if tomorrow I manage to reproduce the crash (or another crash? :) with this fix as well.

**Comment by tigran123** [comment]
Btw, when breaking DjVu I would always have to press `Menu` and `Vol+/-` which does call `renderUtf8Text()`

**Comment by tigran123** [comment]
No, bad news, even with this fix (which is still valid, of course) we do get a crash. Only now I seem to need to press `T` to force it (but it is on files with NO TOC, so the fix by @hwhw should not matter here).

Anyway, the PR #408 is still a valid bugfix, but it seems we still have a serious bug here somewhere. I'll work on it tomorrow, now it is time to sleep :)

**Comment by tigran123** [comment]
Now, the current master tree crashes on ARM with a stacktrace identical to what is shown on the emulator when run via `valgrind`:

```
==14203== Conditional jump or move depends on uninitialised value(s)
==14203==    at 0x4F9CD4: lj_str_new (in /home/tigran/Kindle/kindlepdfviewer/kpdfview.emulator)
==14203==    by 0x4FA1BD: lj_str_fromnum (in /home/tigran/Kindle/kindlepdfviewer/kpdfview.emulator)
==14203==    by 0x5154EC: lj_meta_cat (in /home/tigran/Kindle/kindlepdfviewer/kpdfview.emulator)
==14203==    by 0x50F504: lj_BC_CAT (in /home/tigran/Kindle/kindlepdfviewer/kpdfview.emulator)
==14203==    by 0x4FF35F: lua_pcall (in /home/tigran/Kindle/kindlepdfviewer/kpdfview.emulator)
==14203==    by 0x40765D: main (kpdfview.c:74)
```

So, on the emulator it is something that valgrind detects and on the ARM it causes a SIGSEGV.

**Comment by tigran123** [comment]
Ignore the previous comment. I have narrowed down the actual cause of the crash but I am trying to understand WHY such innocent and looking-right changes (namely my recent changes to rendertext.lua) would cause memory corruption. I suspect that they just triggered the problem that was already there...

**Comment by hwhw** [comment]
Yep, such changes will change memory layouts on stack and heap, thus
leading to situations where crashes either occur loudly or in silence,
without us noticing. My bet is still on memory corruption, probably not at
the places we are looking at currently. These would just be the places
where the error manifest in the end. That's the culprit with memory
corruption and it's what C programmers still earn money with due to the
"power" that language offers...

**Comment by tigran123** [comment]
Yes, now I definitely agree with you that it is a memory corruption. The only question is whether my changes actually _cause_ it or whether they simply _trigger_ it. I am still investigating, will let you know what I find.

**Comment by tigran123** [comment]
Ok, this is the minimal set of changes to rendertex.lua that makes the corruption on ARM go away: SHA1:  edd347eb55

Please review these changes to answer the following question: is this really a fix or does it just hide the real problem?

**Comment by tigran123** [comment]
The reason behind my original changes was this: the `prevcharcode` local variable is created and can never become `nil` (ignoring the 4,5,6-byte UTF-8 case from `util.utf8chardore()`) so the check `if prevcharcode` is meaninless --- it is always true. The correct check for the presence of previous character would be `if prevcharcode ~= 0` because `prevcharcode` was initialized to 0 at the beginning. Otherwise, as things stood, the call to `blitFrom()` was never made and we always went through the kerning code (if `kerning` value passed was `true`).

However, then I discovered that with this correct logic the calls to `blitFrom()` cause the first glyph to be painted with the wrong background. I didn't understand the difference between `blitFrom()` and `addblitFrom()` functions but I observed that simply using `addblitFrom()` produces the correct background.

I violated the rule "never change the code you do not fully understand", i.e. I should have first learned the difference between `blitFrom()` and `addblitFrom()` before substituting one with the other.

**Comment by hwhw** [comment]
addblitFrom does add the pixel values whereas simple blitting will overwrite them. I've added boundary checking to blitbuffer.c now (must be enabled by defining "DEBUG") in that pull request I have sent. Maybe I missed it, but is there a reproducable way to trigger the error on ARM? For which documents does it happen, and under what circumstances?

**Comment by tigran123** [comment]
Yes, it is easy to reproduce. You need this DjVu file:

https://github.com/downloads/tigran123/libdjvu/libdjvu-test.djvu
1. Go to page 15
2. Use `Shift-X` to define a reasonable bbox around the text
3. Press `Next` and do the same on the next page (to have a bbox for both odd and even pages)
4. Press `Prev` to go back to page 15
5. Press `Shift-S` and page `Next` several (5-10) times
6. Sometimes press `Menu` to force rendering some text
7. Sometimes press `Vol+/-` keys to change gamma (and force page redraw). Especially dangerous thing is to change gamma when only the bottom part of the page is visible.
8. Sometimes press `Alt-Next/Prev` to change zoom, always return back to `Shift-S` mode
9. Do a few turns `K/J`

On a crashing build it crashes in less than 2-3 minutes of this activity. On a non-crashing one (with the commit I mentioned above) it never crashes :)

**Comment by hwhw** [comment]
Thank you. That was fast: my new constraints checking errored out on step 2. Investigating...

**Comment by tigran123** [comment]
Yes:

```
violated blitbuffer constraints in file blitbuffer.c, line 338!
```

that's coming from addblitFrom() and the only users thereof are renderUtf8Text() and renderUtf8TextWidth()

**Comment by hwhw** [comment]
the addBlitFrom() was missing some boundary checks. Since they are identical between the BlitFrom()s, I've factored them out into another function.

**Comment by tigran123** [comment]
Ok, I will test your PR now, thank you very much for looking into this! :)

**Comment by tigran123** [comment]
This issue can be closed because @hwhw has fixed the real cause of the corruption.
